### ISSUE TITLE
[macOS] TextField left arrow key while composing deletes text inside composing range

### ISSUE THREAD

**Issue by K-Maron** [flutter/flutter#120852]
## Steps to Reproduce

1. Execute `flutter run` on the code sample 
2. Use Japanese as an input source (Using 
3. **(Don't skip this step)** Type anything and confirm out of its composing state. Make sure the typed letters in this step are not in the composing range. The error doesn't seem to occur if the entire text editing value of the TextField is within the composing range. The idea is to make sure there is at least one letter preceding the error causing composing region. 
4. Type any letter after step 3, and hit space key to bring up a composing candidate list
5. Press the left arrow key, at least 2 times for a single character or more for longer character sequence.
6. Then you should see the text in the composing range disappears. 

### Other notes & concerns
1. I'm not sure if this is an intended behavior. At least other text editors don't behave like this. 
2. Pressing any arrow key while composing makes macOS ring an error sound. It's the sound you often hear when you perform unintended operations. This is actually a much bigger concern. What's going on? Hope it doesn't lead to another issue in the future...

**Expected results:** 
The text inside composing region doesn't disappear with left arrow keys, nor does pressing any arrow key produces any error sound. 

**Actual results:** 
This is me pressing left arrow on my keyboard when text inside composing range disappears.
https://user-images.githubusercontent.com/48747894/219263753-c3a151fd-63e4-4802-b806-9f1fcb9bc7b7.mov

<details>
<summary>flutter doctor -v</summary>
[✓] Flutter (Channel stable, 3.3.0, on macOS 12.6 21G115 darwin-arm, locale en-JP)
    • Flutter version 3.3.0 on channel stable at /Users/kazuki/Developer/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision ffccd96b62 (6 months ago), 2022-08-29 17:28:57 -0700
    • Engine revision 5e9e0e0aa8
    • Dart version 2.18.0
    • DevTools version 2.15.0

[✓] Android toolchain - develop for Android devices (Android SDK version 32.1.0-rc1)
    • Android SDK at /Users/kazuki/Library/Android/sdk
    • Platform android-32, build-tools 32.1.0-rc1
    • Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.13+0-b1751.21-8125866)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.0.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 14A400
    • CocoaPods version 1.11.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2021.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.13+0-b1751.21-8125866)

[✓] Connected device (3 available)
    • iPhone 12 mini (mobile) • 131D4213-85CC-466C-A118-80E33902D7D7 • ios            • com.apple.CoreSimulator.SimRuntime.iOS-16-0 (simulator)
    • macOS (desktop)         • macos                                • darwin-arm64   • macOS 12.6 21G115 darwin-arm
    • Chrome (web)            • chrome                               • web-javascript • Google Chrome 110.0.5481.100

[✓] HTTP Host Availability
    • All required HTTP hosts are available

• No issues found!

</details>

<details>
<summary>Code sample</summary>

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return const MaterialApp(
      home: MyHomePage(),
    );
  }
}

class MyHomePage extends StatelessWidget {
  const MyHomePage({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return const Scaffold(
      body: Center(
        child: Padding(
          padding: EdgeInsets.all(20),
          child: TextField(),
        ),
      ),
    );
  }
}

```

</details>

<details>
  <summary>Logs</summary>

Can't find any error logs being printed...

</details>

**Comment by darshankawar** [comment]
Thanks for the report @K-Maron 
Is the behavior only occuring on desktop platform, or on mobile too (ex: ios simulator) and on web ?

I used your code sample and ran it on macOS desktop platform using latest master version of the framework.
I first selected Japanese Kana language and entered `test` word and followed your steps to replicate. Tried left arrow key which didn't close the composing range disappear.
I then tried with `abc`, hit space and then left arrow key with which I probably saw the composing range disappear.

I am not completely sure if this is the correct interpretation or representation of the behavior you reported, but I observed as shown in below video:


https://user-images.githubusercontent.com/67046386/219316795-8b4cf349-d996-4c93-ac53-e105399fb88c.mov



> * **(Don't skip this step)** Type anything and confirm out of its composing state. Make sure the typed letters in this step are not in the composing range. The error doesn't seem to occur if the entire text editing value of the TextField is within the composing range. The idea is to make sure there is at least one letter preceding the error causing composing region.
> * Type any letter after step 3, and hit space key to bring up a composing candidate list

If above is not completely correct that represents your issue, please provide sample text that we can enter as-is (using English keyboard) so that we can use it to verify this properly.

Also, I see that you are on `3.3.10`, so I also suggest you to run your scenario on latest master or stable to check if the behavior still persists on them or not.

**Comment by K-Maron** [comment]
> (Don't skip this step) Type anything and confirm out of its composing state. Make sure the typed letters in this step are not in the composing range. The error doesn't seem to occur if the entire text editing value of the TextField is within the composing range. The idea is to make sure there is at least one letter preceding the error causing composing region.

From the video you recorded, I can see that you are actually skipping this step above. Composing range should not take up the whole text of the TextField in order for this issue to be reproduced. There has to be other text preceding the error causing composing range. 

To illustrate my point further, for example, type keyboard "A" to produce "あ" and hit enter to exit composing mode for this character. Then type keyboard "U" to produce "う" and this time press left arrow key at least twice or more while in composing and see if the error happens. As I said the preceding "あ" here is needed to be out of the error causing composing range "い". 

To further clarify my point above, please check this [flutter api doc](https://api.flutter.dev/flutter/widgets/TextEditingController/buildTextSpan.html) to find an implementation of buildTextSpan method in TextEditingController class.
```dart
TextSpan(text: value.composing.textBefore(value.text)),
TextSpan(
  style: composingStyle,
  text: value.composing.textInside(value.text),
),
TextSpan(text: value.composing.textAfter(value.text)),
```

What I mean by saying preceding letters need to exist for this error to reproduce is that value.composing.textBefore(value.text) should not be empty! So in the example provided above, value.composing.textBefore(value.text) = "あ" and value.composing.textInside(value.text) = "い". In your video, you have value.composing.textBefore(value.text) = "" as empty string. By this, you won't be able to reproduce the error.  Hope I'm clear here.

Regarding flutter version upgrade, I checked the release note of 3.7.0 and change logs before posing this issue. It doesn't seem like anything related to this issue is addressed. I'll see if I can find time to go through the upgrade, but please check on your side as well if you can reproduce the error based on the further explanation above. Steps to reproduce provided initially is still correct it seems to me at least... 

Lastly, I'm also using US keyboard ;)

**Comment by K-Maron** [comment]
Oh sorry, forgot to mention. @darshankawar I've only observed this behavior on macOS. Haven't tested on any other platform.

**Comment by K-Maron** [comment]
@darshankawar I just upgraded and check with 3.7.3 and confirmed that it only made things worse. Now it removes not only the composing region but also the preceding letters... 

```
[✓] Flutter (Channel stable, 3.7.3, on macOS 12.6 21G115 darwin-arm64, locale en-JP)
    • Flutter version 3.7.3 on channel stable at /Users/kazuki/Developer/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 9944297138 (8 days ago), 2023-02-08 15:46:04 -0800
    • Engine revision 248290d6d5
    • Dart version 2.19.2
    • DevTools version 2.20.1
```

**Comment by K-Maron** [comment]
Recorded the new behavior with 3.7.3 on mac.

https://user-images.githubusercontent.com/48747894/219529473-17be99cd-02a2-49d7-82b9-368cd3884552.mov

**Comment by darshankawar** [comment]
Thanks for the update. I verified on other platforms (mobile and web) and it doesn't seem to be occuring on those and seems to be specific to desktop.

1. web:


https://user-images.githubusercontent.com/67046386/219562679-9f55d45b-c74a-46e3-a043-331757ce036a.mov


2. mobile:


https://user-images.githubusercontent.com/67046386/219562714-7e11ecfc-2644-4905-9362-5b301810fd31.mov


3. desktop (occuring):


https://user-images.githubusercontent.com/67046386/219562741-3334b959-8965-40ca-b00a-2aa8dc8fe4eb.mov

<details>
<summary> stable, master flutter doctor -v </summary>

```
[!] Flutter (Channel stable, 3.7.3, on macOS 12.2.1 21D62 darwin-x64, locale
    en-GB)
    • Flutter version 3.7.3 on channel stable at
      /Users/dhs/documents/fluttersdk/flutter
    ! Warning: `flutter` on your path resolves to
      /Users/dhs/Documents/Fluttersdk/flutter/bin/flutter, which is not inside
      your current Flutter SDK checkout at
      /Users/dhs/documents/fluttersdk/flutter. Consider adding
      /Users/dhs/documents/fluttersdk/flutter/bin to the front of your path.
    ! Warning: `dart` on your path resolves to
      /Users/dhs/Documents/Fluttersdk/flutter/bin/dart, which is not inside your
      current Flutter SDK checkout at /Users/dhs/documents/fluttersdk/flutter.
      Consider adding /Users/dhs/documents/fluttersdk/flutter/bin to the front
      of your path.
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 9944297138 (4 days ago), 2023-02-08 15:46:04 -0800
    • Engine revision 248290d6d5
    • Dart version 2.19.2
    • DevTools version 2.20.1
    • If those were intentional, you can disregard the above warnings; however
      it is recommended to use "git" directly to perform update checks and
      upgrades.

[!] Xcode - develop for iOS and macOS (Xcode 12.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    ! Flutter recommends a minimum Xcode version of 13.
      Download the latest version or update via the Mac App Store.
    • CocoaPods version 1.11.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] VS Code (version 1.62.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.21.0

[✓] Connected device (5 available)
    • SM G975F (mobile)       • RZ8M802WY0X • android-arm64   • Android 11 (API 30)
    • Darshan's iphone (mobile)  • 21150b119064aecc249dfcfe05e259197461ce23 •
      ios            • iOS 14.4.1 18D61
    • iPhone 12 Pro Max (mobile) • A5473606-0213-4FD8-BA16-553433949729     •
      ios            • com.apple.CoreSimulator.SimRuntime.iOS-14-3 (simulator)
    • macOS (desktop)            • macos                                    •
      darwin-x64     • Mac OS X 10.15.4 19E2269 darwin-x64
    • Chrome (web)               • chrome                                   •
      web-javascript • Google Chrome 98.0.4758.80

[✓] HTTP Host Availability
    • All required HTTP hosts are available

! Doctor found issues in 1 category.

[!] Flutter (Channel master, 3.8.0-13.0.pre.52, on macOS 12.2.1 21D62
    darwin-x64, locale en-GB)
    • Flutter version 3.8.0-13.0.pre.52 on channel master at
      /Users/dhs/documents/fluttersdk/flutter
    ! Warning: `flutter` on your path resolves to
      /Users/dhs/Documents/Fluttersdk/flutter/bin/flutter, which is not inside
      your current Flutter SDK checkout at
      /Users/dhs/documents/fluttersdk/flutter. Consider adding
      /Users/dhs/documents/fluttersdk/flutter/bin to the front of your path.
    ! Warning: `dart` on your path resolves to
      /Users/dhs/Documents/Fluttersdk/flutter/bin/dart, which is not inside your
      current Flutter SDK checkout at /Users/dhs/documents/fluttersdk/flutter.
      Consider adding /Users/dhs/documents/fluttersdk/flutter/bin to the front
      of your path.
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision cc473e4f19 (73 minutes ago), 2023-02-15 22:51:19 -0500
    • Engine revision 3fc40ca5be
    • Dart version 3.0.0 (build 3.0.0-241.0.dev)
    • DevTools version 2.21.1
    • If those were intentional, you can disregard the above warnings; however
      it is recommended to use "git" directly to perform update checks and
      upgrades.
      
[!] Xcode - develop for iOS and macOS (Xcode 12.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    ! Flutter recommends a minimum Xcode version of 13.
      Download the latest version or update via the Mac App Store.
    • CocoaPods version 1.11.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] VS Code (version 1.62.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.21.0

[✓] Connected device (5 available)
    • SM G975F (mobile)       • RZ8M802WY0X • android-arm64   • Android 11 (API 30)
    • Darshan's iphone (mobile)  • 21150b119064aecc249dfcfe05e259197461ce23 •
      ios            • iOS 14.4.1 18D61
    • iPhone 12 Pro Max (mobile) • A5473606-0213-4FD8-BA16-553433949729     •
      ios            • com.apple.CoreSimulator.SimRuntime.iOS-14-3 (simulator)
    • macOS (desktop)            • macos                                    •
      darwin-x64     • Mac OS X 10.15.4 19E2269 darwin-x64
    • Chrome (web)               • chrome                                   •
      web-javascript • Google Chrome 98.0.4758.80

[✓] HTTP Host Availability
    • All required HTTP hosts are available

! Doctor found issues in 1 category.



```

</details>

**Comment by K-Maron** [comment]
@darshankawar Thanks for checking into the details. I saw the recording of the web version. It doesn't seem like a correct behavior on web either. It is taking the text before again into the composing region.

**Comment by gnprice** [comment]
> 2\. Pressing any arrow key while composing makes macOS ring an error sound. It's the sound you often hear when you perform unintended operations. This is actually a much bigger concern. What's going on? Hope it doesn't lead to another issue in the future...

@K-Maron Would you file a separate issue for this other behavior? That will help ensure both of these useful issue reports can get properly tracked and addressed.

**Comment by justinmc** [comment]
CC @LongCatIsLooong from triage.
